rust/lance-encoding/src/data.rs

```diff
@@ -372,17 +307,851 @@ impl DataBlock for DictionaryDataBlock {
+        match self {
+            Self::AllNull(inner) => inner.num_values,
+            Self::Nullable(inner) => inner.data.num_values(),
+            Self::FixedWidth(inner) => inner.num_values,
+            Self::VariableWidth(inner) => inner.num_values,
+            Self::Struct(inner) => inner.children[0].num_values(),
+            Self::Dictionary(inner) => inner.indices.num_values,
+        }
+    }
+}
+
+macro_rules! as_type {
+    ($fn_name:ident, $inner:tt, $inner_type:ident) => {
+        pub fn $fn_name(self) -> Result<$inner_type> {
+            match self {
+                Self::$inner(inner) => Ok(inner),
+                _ => Err(Error::Internal {
+                    message: format!("Expected {}, got {}", stringify!($inner), self.name()),
+                    location: location!(),
+                }),
+            }
+        }
+    };
+}
+
+// Cast implementations
+impl DataBlock {
+    as_type!(as_all_null, AllNull, AllNullDataBlock);
+    as_type!(as_nullable, Nullable, NullableDataBlock);
+    as_type!(as_fixed_width, FixedWidth, FixedWidthDataBlock);
+    as_type!(as_variable_width, VariableWidth, VariableWidthBlock);
+    as_type!(as_struct, Struct, StructDataBlock);
+    as_type!(as_dictionary, Dictionary, DictionaryDataBlock);
+}
+
+// Methods to convert from Arrow -> DataBlock
+
+fn get_byte_range<T: ArrowNativeType>(offsets: &mut LanceBuffer) -> Range<usize> {
+    let offsets = offsets.borrow_to_typed_slice::<T>();
+    if offsets.as_ref().is_empty() {
+        0..0
+    } else {
+        offsets.as_ref().first().unwrap().as_usize()..offsets.as_ref().last().unwrap().as_usize()
+    }
+}
+
+// Given multiple offsets arrays [0, 5, 10], [0, 3, 7], etc. stitch
+// them together to get [0, 5, 10, 13, 20, ...]
+//
+// Also returns the data range referenced by each offset array (may
+// not be 0..len if there is slicing involved)
+fn stitch_offsets<T: ArrowNativeType + std::ops::Add<Output = T> + std::ops::Sub<Output = T>>(
+    offsets: Vec<LanceBuffer>,
+) -> (LanceBuffer, Vec<Range<usize>>) {
+    if offsets.is_empty() {
+        return (LanceBuffer::empty(), Vec::default());
+    }
+    let len = offsets.iter().map(|b| b.len()).sum::<usize>();
+    // Note: we are making a copy here, even if there is only one input, because we want to
+    // normalize that input if it doesn't start with zero.  This could be micro-optimized out
+    // if needed.
+    let mut dest = Vec::with_capacity(len);
+    let mut byte_ranges = Vec::with_capacity(offsets.len());
+
+    dest.push(T::from_usize(0).unwrap());
+    for mut o in offsets.into_iter() {
+        if !o.is_empty() {
+            let last_offset = *dest.last().unwrap();
+            let o = o.borrow_to_typed_slice::<T>();
+            let start = *o.as_ref().first().unwrap();
+            dest.extend(
+                // Skip the leading zero, add last_offset, subtract start
+                o.as_ref()[1..].iter().map(|&x| x + last_offset - start),
+            );
```

Comment:
here we are assuming every offsets `LanceBuffer` starts with 0, do we have logic to enforce it somewhere?
one caveat for example: 
arrow offsets array a: [0, 5, 10, 13, 20, ...]
let a_slice = a.slice(1, 4), then `a_slice` is a offsets array of [5, 10, 113, 20]
if we can convert this arrow offsets array directly without adjustment,
then pass it in here, the offsets here then is wrong

Reply:
> do we have logic to enforce it somewhere?

That's part of what we're doing here.  In Arrow, buffers can be sliced, and string offsets might not start with zero.  At the moment, during conversion from Arrow, I'm doing a lot of normalization.  This isn't free, but writer performance typically isn't critical and it will make it easier to build encoders.  We can relax these checks in the future as we need to optimize writers.

So the output here should be an offsets buffer that starts with zero.  There are two things we are doing here.  First, we are adding the current offsets buffer to the end of the previous offsets buffer (`last_offset`).  Second, we are normalizing any non-zero start value (`start`).

What you're asking about should be addressed by `start`.  This is initialized to `*o.as_ref().first().unwrap()` (the first offset value).  So, in your example, `start = 5`.  Since your example has only one input array then `last_offset` will be `0`.

So for each value we are calculating `x + 0 - 5`.  Given `[5, 10, 13, 20]` the output will be `[0, 5, 8, 15]`.

Reply:
gotcha, yes, I see it now.
```
// Skip the leading zero, add last_offset, subtract start
```
this comment is confusing though, `the leading zero`

Reply:
Good point.  I've expanded on the comment, hopefully clarifying.